File: 2121-find-if-path-exists-in-graph/2121-find-if-path-exists-in-graph.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        graph = [[] for _ in range(n)]
        visited = [False for _ in range(n)]
        for i in range(len(edges)):
            edge = edges[i]
            graph[edge[0]].append(edge[1])
            graph[edge[1]].append(edge[0])

        self.dfs(graph, source, destination, visited)
        return visited[destination]

    def dfs(self, graph: List[List[int]], v: int, target: int, visited: List[bool]) -> bool:
        visited[v] = True
        for i in graph[v]:
            if i == target:
                visited[i] = True
                break
            if not visited[i]:
                self.dfs(graph, i, target, visited)

File: 2121-find-if-path-exists-in-graph/2121-find-if-path-exists-in-graph.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public boolean validPath(int n, int[][] edges, int source, int destination) {
        List<List<Integer>> graph = new ArrayList<>();
        boolean[] visited = new boolean[n];

        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }

        for (int[] edge : edges) {
            graph.get(edge[0]).add(edge[1]);
            graph.get(edge[1]).add(edge[0]);
        }

        dfs(graph, source, destination, visited);
        return visited[destination];
    }

    private void dfs(List<List<Integer>> graph, int v, int target, boolean[] visited) {
        visited[v] = true;

        for (int neighbor : graph.get(v)) {
            if (neighbor == target) {
                visited[neighbor] = true;
                break;
            }

            if (!visited[neighbor]) {
                dfs(graph, neighbor, target, visited);
            }
        }
    }
}
